src/Servers.jl

```diff
@@ -130,278 +109,196 @@ function check_rate_limit(tcp;
+            let io=io, count=count
+                @async begin
+                    try
+                        verbose && @info "Accept ($count):  $conn"
+                        handle(h, conn, reuse_limit, readtimeout)
+                        verbose && @info "Closed ($count):  $conn"
                     catch e
-                        @error "Error ($i):  $conn" exception=(e, stacktrace(catch_backtrace()))
+                        @error exception=(e, stacktrace(catch_backtrace()))
                     finally
                         connectioncounter[] -= 1
-                        close(conn)
-                        @info "Closed ($i):  $conn"
+                        close(io)
+                        verbose && @info "Closed ($count):  $conn"
                     end
                 end
-            catch e
-                if e isa Base.IOError
-                    @warn "$e"
-                    break
-                else
-                    rethrow(e)
-                end
             end
-            id += 1
-        end
-    catch e
-        if e isa InterruptException
-            @warn "Interrupted: listen($hostname)"
-        else
-            rethrow(e)
+        catch e
+            if e isa InterruptException
+                @warn "Interrupted: listen($server)"
+                close(server)
+                break
+            end
+            @error exception=(e, stacktrace(catch_backtrace()))
         end
-    finally
-        close(tcpserver)
+        count += 1
     end
-
     return
 end
 
-
 """
-Start a timeout monitor task to close the `Connection` if it is inactive.
-Create a `Transaction` object for each HTTP Request received.
+Connection handler: starts an async readtimeout thread if needed, then creates
+Transactions to be handled as long as the Connection stays open. Only reuse_limit + 1
+# of Transactions will be allowed during the lifetime of the Connection.
 """
-function handle_connection(f::Function, c::Connection;
-                           reuse_limit::Int=nolimit,
-                           readtimeout::Int=0, kw...)
+function handle(h, c::Connection,
+                reuse_limit::Int=10,
+                readtimeout::Int=0)
 
     wait_for_timeout = Ref{Bool}(true)
-    if readtimeout > 0
-        @async while wait_for_timeout[]
-            @show inactiveseconds(c)
-            if inactiveseconds(c) > readtimeout
-                @warn "Timeout: $c"
-                writeheaders(c.io, Response(408, ["Connection" => "close"]))
-                close(c)
-                break
-            end
-            sleep(8 + rand() * 4)
-        end
-    end
-
+    readtimeout > 0 && check_readtimeout(c, readtimeout, wait_for_timeout)
     try
         count = 0
         while isopen(c)
-            io = Transaction(c)
-            handle_transaction(f, io; final_transaction=(count == reuse_limit), kw...)
+            handle(h, Transaction(c), count == reuse_limit)
```

Comment:
Better to keep this as a kw arg. Positional bool arguments are awesomely hard to code review. Writing it out as `final_transaction=(count == reuse_limit)` makes it clear.